Check prime for `Pow` and `Mul` classes.

AadityaNair opened the pull request:
Added an `_eval_is_prime()` function for `Pow` and `Mul` classes.
Closes #8488 and #9151

AadityaNair committed:
Added `is_prime` for `Mul` class.

Closes sympy/sympy#8488.

AadityaNair committed:
Added tests checking `is_prime` for Mul class
Improved `is_prime` for Mul

AadityaNair committed:
Added  `is_prime` functionality `Pow` class.

AadityaNair committed:
Added tests for `is_prime` functionality in `Pow` class.

AadityaNair committed:
Better `is_prime` handling.

debugger22 commented on sympy/core/tests/test_assumptions.py:
```diff
@@ -714,6 +714,35 @@ def test_Pow_is_algebraic():
+    from sympy import Mul
+    x = Symbol('x', positive=True, integer=True)
+    y = Symbol('y', positive=True, integer=True)
+    assert (x*y).is_prime is False
```
Either of `x` or `y` can attain `1` according to provided facts. So, I think, it shouldn't return `False`.

debugger22 commented on sympy/core/tests/test_assumptions.py:
```diff
@@ -714,6 +714,35 @@ def test_Pow_is_algebraic():
+    from sympy import Pow
+    x = Symbol('x', positive=True, integer=True)
+    y = Symbol('y', positive=True, integer=True)
+    assert (x**y).is_prime is False
```
Same goes for this.

AadityaNair committed:
Updated code based on suggestions.

Earlier:
```
>>> x = Symbol('x', positive=True, integer=True)
>>> y = Symbol('y', positive=True, integer=True)
>>> (x*y).is_prime
False
```

Now
```
>>> x = Symbol('x', positive=True, integer=True)
>>> y = Symbol('y', positive=True, integer=True)
>>> (x*y).is_prime
None
```

AadityaNair committed:
Updated test.

AadityaNair commented:
Updated the code.
Now,
```python
>>> x = Symbol('x', positive=True, integer=True)
>>> y = Symbol('y', positive=True, integer=True)
>>> (x*y).is_prime 
None
```

debugger22 commented on sympy/core/mul.py:
```diff
@@ -1182,6 +1182,43 @@ def _eval_is_even(self):
+            If there are more than one of such a symbol then the result is not prime.
+            Else, the result cannot be determined.
+            """
+            from sympy.functions.elementary.miscellaneous import Min
```
Where did you use `Min`?

AadityaNair replied:
Sorry, was using it to determine the minimum value of symbol.
Forgot I removed it.

debugger22 commented on sympy/core/mul.py:
```diff
@@ -1182,6 +1182,43 @@ def _eval_is_even(self):
+            Else, the result cannot be determined.
+            """
+            from sympy.functions.elementary.miscellaneous import Min
+            number_of_args=0 # count of symbols with minimum value greater than one
```
Space around `=`.

debugger22 commented on sympy/core/mul.py:
```diff
@@ -1182,6 +1182,43 @@ def _eval_is_even(self):
+                except TypeError: # If it cannot be determined.
+                    pass
+
+            if number_of_args >1:
```
Space around `>`.

AadityaNair committed:
Updated code quality.

debugger22 commented on sympy/core/tests/test_assumptions.py:
```diff
@@ -714,6 +714,36 @@ def test_Pow_is_algebraic():
+    assert (x**y).is_prime is None
+
+    x = Symbol('x', prime=True)
+    assert (x**1).is_prime is True
```
In master:
```
In [321]: x = Symbol('x', prime=True)

In [322]: (x**1).is_prime
Out[322]: True
```

AadityaNair replied:
I do not see the problem.

debugger22 replied:
> I do not see the problem.

It's already in the master.

AadityaNair replied:
there is no `_eval_is_prime` implemented in master.

AadityaNair replied:
You are seeing that because `x**1` is directly evaluated to `x`. So the line is essentially
`x.is_prime` which is `True`.

debugger22 replied:
> You are seeing that because x**1 is directly evaluated to x. So the line is essentially
x.is_prime which is True.

Same is happening in your branch.
```
In [5]: type(x**1)
Out[5]: sympy.core.symbol.Symbol
```

AadityaNair replied:
Should I remove it or leave it there ?

debugger22 replied:
I think you should remove it because it is not testing what it is intended for.

debugger22 commented on sympy/core/tests/test_assumptions.py:
```diff
@@ -714,6 +714,36 @@ def test_Pow_is_algebraic():
+    assert (x**y).is_prime is None
+
+    x = Symbol('x', positive=True)
+    assert (x**y).is_prime is None
```
In master:
```
In [323]: x = Symbol('x', positive=True)

In [324]: y = Symbol('y', positive=True, integer=True)

In [325]: (x**y).is_prime
```

AadityaNair replied:
Since no `_eval_is_prime` is implemented, `(x**y).is_prime` returns `None` as with all the assumptions that are not implemented.

debugger22 replied:
Oh. If this test is not in the master, you can ignore this.

AadityaNair committed:
Removed unnecessary tests.

AadityaNair committed:
Better checking of minimum values.

Initially `x>1` was used to check whether x is greater than 1.
Now `(x-1).is_positive` is used.

AadityaNair commented:
Updated the code with better way of checking `greater than 1`

asmeurer commented on sympy/core/power.py:
```diff
@@ -487,6 +487,20 @@ def _eval_is_finite(self):
+            a Power will be non-prime only if both base and exponent
+            are greater than 1
+            """
+            if (self.base-1).is_positive or (self.exp-1).is_positive:
```
There need to be some checks if they are integers as well. For example, `E > 1` and `log(3) > 1`, but `exp(log(3)) == 3` is prime.

AadityaNair replied:
> `exp(log(3)) == 3` is prime.

`exp(log(3))` is an instance of `sympy.functions.elementary.exponential.exp`.
So nothing really can be done here.

Use of any special functions will lead to an object of its own class.
Using explicit values result in a `Float` number which cannot be evaluated as prime.

> There need to be some checks if they are integers as well.

Control reaches here only if the power can be evaluated as an integer

certik replied:
I think @AadityaNair is right.

certik commented:
That looks good to me.
